src/Sylius/Bundle/CoreBundle/spec/Validator/Constraints/CatalogPromotionRulesValidityValidatorSpec.php

```diff
@@ -0,0 +1,93 @@
+
+namespace spec\Sylius\Bundle\CoreBundle\Validator\Constraints;
+
+use PhpSpec\ObjectBehavior;
+use Sylius\Bundle\CoreBundle\Validator\Constraints\CatalogPromotionRulesValidity;
+use Sylius\Component\Core\Model\ProductVariantInterface;
+use Sylius\Component\Core\Repository\ProductVariantRepositoryInterface;
+use Sylius\Component\Promotion\Model\CatalogPromotionRuleInterface;
+use Symfony\Component\Validator\Constraint;
+use Symfony\Component\Validator\ConstraintValidator;
+use Symfony\Component\Validator\Context\ExecutionContextInterface;
+use Symfony\Component\Validator\Violation\ConstraintViolationBuilderInterface;
+
+final class CatalogPromotionRulesValidityValidatorSpec extends ObjectBehavior
+{
+    function let(ExecutionContextInterface $executionContext, ProductVariantRepositoryInterface $variantRepository): void
+    {
+        $this->beConstructedWith($variantRepository);
+
+        $this->initialize($executionContext);
+    }
+
+    function it_is_a_constraint_validator(): void
+    {
+        $this->shouldHaveType(ConstraintValidator::class);
+    }
+
+    function it_adds_violation_if_catalog_promotion_has_invalid_type(
+        ExecutionContextInterface $executionContext,
+        ConstraintViolationBuilderInterface $constraintViolationBuilder,
+        CatalogPromotionRuleInterface $rule
+    ): void {
+        $constraint = new CatalogPromotionRulesValidity();
+
+        $rule->getType()->willReturn('wrong_type');
+
+        $executionContext->addViolation($constraint->invalid_type)->willReturn($constraintViolationBuilder)->shouldBeCalled();
+
+        $this->validate([$rule], $constraint);
+    }
+
+    function it_adds_violation_if_catalog_promotion_rule_is_invalid(
+        ExecutionContextInterface $executionContext,
+        ConstraintViolationBuilderInterface $constraintViolationBuilder,
+        CatalogPromotionRuleInterface $rule
+    ): void {
+        $constraint = new CatalogPromotionRulesValidity();
+
+        $rule->getType()->willReturn(CatalogPromotionRuleInterface::TYPE_CONTAINS_VARIANTS);
+        $rule->getConfiguration()->willReturn(['wrong_config' => 'wrong_value']);
+
+        $executionContext->addViolation($constraint->invalid_configuration)->willReturn($constraintViolationBuilder)->shouldBeCalled();
+
+        $this->validate([$rule], $constraint);
+    }
+
+    function it_does_nothing_if_catalog_promotion_rule_is_valid(
+        ExecutionContextInterface $executionContext,
+        ConstraintViolationBuilderInterface $constraintViolationBuilder,
+        CatalogPromotionRuleInterface $rule,
+        ProductVariantRepositoryInterface $variantRepository,
+        ProductVariantInterface $firstVariant,
+        ProductVariantInterface $secondVariant
+    ): void {
+        $constraint = new CatalogPromotionRulesValidity();
+
+        $rule->getType()->willReturn(CatalogPromotionRuleInterface::TYPE_CONTAINS_VARIANTS);
+        $rule->getConfiguration()->willReturn(['fir_var', 'sec_var']);
```

Comment:
Tell me did you do in this time, that you saved writing `fir_var` instead of `first_variant` 😄

Reply:
Maybe he meant 'fir' as a type of tree 🌲